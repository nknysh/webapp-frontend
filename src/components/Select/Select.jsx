import React from 'react';
import { mapObjIndexed, values } from 'ramda';

import { propTypes, defaultProps } from './Select.props';
import { StyledSelect, StyledFormLabel, MaterialSelect, SelectMenuItem, selectClasses } from './Select.styles';

const renderItems = (value, key) => (
  <SelectMenuItem key={key} value={key}>
    {value}
  </SelectMenuItem>
);

export const Select = ({ label, onSelected, options, className, onChange, ...props }) => {
  const onSelect = e => {
    onSelected(e.target.value);
    onChange(e);
  };

  return (
    <StyledSelect className={className}>
      <StyledFormLabel
        control={
          <MaterialSelect classes={selectClasses} onChange={onSelect} {...props}>
            {values(mapObjIndexed(renderItems, options))}
          </MaterialSelect>
        }
        label={label}
      />
    </StyledSelect>
  );
};

Select.propTypes = propTypes;
Select.defaultProps = defaultProps;

export default Select;
