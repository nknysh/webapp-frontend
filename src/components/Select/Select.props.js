import PropTypes from 'prop-types';

import { noop } from 'utils';

export const propTypes = {
  label: PropTypes.string,
  onSelected: PropTypes.func,
  options: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
};

export const defaultProps = {
  onSelected: noop,
  options: {},
};
