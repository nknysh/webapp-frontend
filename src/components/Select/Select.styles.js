import styled from 'styled-components';
import { Select, FormControlLabel, MenuItem } from '@material-ui/core';

import theme from 'styles/theme';
import { inputStyles, Hr } from 'styles/elements';
import breakpoints from 'styles/breakpoints';

const baseClass = 'material';

export const selectClasses = {
  root: `${baseClass}-select`,
  select: `${baseClass}-select__input`,
};

export const StyledFormLabel = styled(FormControlLabel)`
  margin: 0 !important;
  width: 100%;

  ${breakpoints.tablet`
    width: auto;
  `}
`;

export const StyledSelect = styled.div`
  label,
  label span {
    font-family: ${theme.defaultFont} !important;
  }
`;

export const MaterialSelect = styled(Select)`
  margin-bottom: ${theme.gutter * 2}px;
  width: 100%;
  font-size: 14px !important;
  
  :before, :after {
    content: '' !important;
    border: 0 !important;
  }

  .${baseClass}-select {
    min-width: 195px;
    width: 100%;

    ${breakpoints.tablet`
      width: auto;
    `}

    &__input {
      ${inputStyles}
      padding: ${theme.gutter}px;
    }
  }
`;

export const SelectMenuItem = styled(MenuItem)``;

export const SectionDivider = styled(Hr)``;
