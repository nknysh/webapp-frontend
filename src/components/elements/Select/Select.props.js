import PropTypes from 'prop-types';

import { noop } from 'utils';

export const propTypes = {
  label: PropTypes.string,
  onChange: PropTypes.func,
  onClose: PropTypes.func,
  onOpen: PropTypes.func,
  onSelected: PropTypes.func,
  open: PropTypes.bool,
  options: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
};

export const defaultProps = {
  onChange: noop,
  onClose: noop,
  onOpen: noop,
  onSelected: noop,
  open: false,
  options: {},
};
