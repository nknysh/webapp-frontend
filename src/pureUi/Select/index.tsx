import React from 'react';
import { IValueLabelPair } from '../../interfaces';
import styled from 'styled-components';

const StyledSelect = styled.select`
  width: 100%;
  font-family: HurmeGeometricSans2, 'Open Sans', sans-serif !important;
  padding: 10px;
  text-transform: uppercase;
  display: block;
  border: 1px solid #e0e0e0;
  color: #373c46;
  width: 100%;
  box-sizing: border-box;
  background-color: white;
  appearance: none;
  border-radius: 0;
`;

const Select = ({
  value,
  options,
  onChange,
  className,
}: {
  value: string;
  options: IValueLabelPair[];
  onChange: Function;
  className?: string;
}) => {
  return (
    <StyledSelect
      className={className || ''}
      value={value}
      onChange={e => {
        if (e.target.value === 'null') {
          return onChange({
            target: {
              value: null,
            },
          });
        }
        return onChange(e);
      }}
    >
      {options &&
        options.map(option => {
          if (option.value === null) {
            return (
              <option key={`${option.value}-${option.label}`} value={'null'} disabled={option.disabled}>
                {option.label}
              </option>
            );
          }
          return (
            <option key={`${option.value}-${option.label}`} value={option.value} disabled={option.disabled}>
              {option.label}
            </option>
          );
        })}
    </StyledSelect>
  );
};

export default Select;
