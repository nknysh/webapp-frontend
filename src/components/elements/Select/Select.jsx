import React, { useCallback, useState } from 'react';
import { append, mapObjIndexed, values, map, pipe } from 'ramda';

import Hr from 'components/elements/Hr';
import { useBodyClass } from 'effects';
import { isArray } from 'utils';

import { propTypes, defaultProps } from './Select.props';
import { MaterialSelect, selectClasses, SelectMenuItem, StyledFormLabel, StyledSelect } from './Select.styles';

const renderItem = (value, key) => (
  <SelectMenuItem key={key} value={key}>
    {value}
  </SelectMenuItem>
);

const renderItems = mapObjIndexed(renderItem);

const mapOverKeys = pipe(
  renderItems,
  values
);

const renderSection = pipe(
  mapOverKeys,
  append(<Hr />)
);

const mapOverSections = map(renderSection);

export const Select = ({ label, onSelected, options, className, onChange, onOpen, onClose, open, ...props }) => {
  const [isOpen, setIsOpen] = useState(open);
  const { setBodyClass, unsetBodyClass } = useBodyClass();

  const onSelect = useCallback(
    e => {
      onSelected(e.target.value);
      onChange(e);
    },
    [onChange, onSelected]
  );

  const onSelectOpen = useCallback(
    (...args) => {
      onOpen(...args);
      setIsOpen(true);
      setBodyClass('select-open');
    },
    [onOpen, setBodyClass]
  );

  const onSelectClose = useCallback(
    (...args) => {
      onClose(...args);
      setIsOpen(false);
      unsetBodyClass('select-open');
    },
    [onClose, unsetBodyClass]
  );

  const renderedOptions = isArray(options) ? mapOverSections(options) : mapOverKeys(options);

  return (
    <StyledSelect className={className}>
      <StyledFormLabel
        control={
          <MaterialSelect
            classes={selectClasses}
            onChange={onSelect}
            onClose={onSelectClose}
            onOpen={onSelectOpen}
            open={isOpen}
            {...props}
          >
            {renderedOptions}
          </MaterialSelect>
        }
        label={label}
      />
    </StyledSelect>
  );
};

Select.propTypes = propTypes;
Select.defaultProps = defaultProps;

export default Select;
