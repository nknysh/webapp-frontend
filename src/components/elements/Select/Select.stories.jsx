import React from 'react';

import { storiesOf } from '@storybook/react';
import { withKnobs } from '@storybook/addon-knobs';

import Select from './Select';

storiesOf('Select', module)
  .addDecorator(withKnobs)
  .add('default', () => {
    return <Select />;
  });
