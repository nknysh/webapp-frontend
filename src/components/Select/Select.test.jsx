import React from 'react';

import { Select } from './Select';

const getComponent = props => shallow(<Select {...props} />);

describe('<Select />', () => {
  describe('render', () => {
    it('matches snapshot', () => {
      expect(getComponent()).toMatchSnapshot();
    });
  });
});
