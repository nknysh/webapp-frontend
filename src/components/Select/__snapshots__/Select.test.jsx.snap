// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> render matches snapshot 1`] = `
<styled.div>
  <Styled(WithStyles(FormControlLabel))
    control={
      <ForwardRef
        classes={
          Object {
            "root": "material-select",
            "select": "material-select__input",
          }
        }
        onChange={[Function]}
      />
    }
  />
</styled.div>
`;
